scripts/circleci_scraper/scraper.py

```diff
@@ -0,0 +1,348 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""CircleCIScraper and related objects"""
+
+import json
+import logging
+import os
+from datetime import datetime, timezone
+from typing import Any
+
+import requests
+from requests.exceptions import RequestException
+
+from client import (
+    CircleCIClient,
+    PipelineGroup,
+    WorkflowGroup,
+    JobGroup,
+    Job,
+    TestMetadataGroup,
+    ArtifactGroup,
+    Artifact,
+    Workflow,
+)
+from config import CircleCIScraperPipelineConfig, MainConfig
+
+
+class CircleCIScraperError(Exception):
+    """Custom exception class for CircleCIScraper errors."""
+
+    def __init__(self, message: str, inner_exception: Exception | None = None) -> None:
+        """Initialize the CircleCIScraperError.
+
+        Args:
+            message (str): The error message.
+            inner_exception (Exception | None): The inner exception that caused this error.
+                                                Defaults to None.
+        """
+        super().__init__(message)
+        self.inner_exception = inner_exception
+
+    def __str__(self) -> str:
+        """Return the string representation of the error.
+
+        Returns:
+            str: The string representation of the error, including the inner exception if present.
+        """
+        if self.inner_exception:
+            return f"{super().__str__()} (caused by {self.inner_exception})"
+        return super().__str__()
+
+
+class CircleCIScraper:
+    """Export CircleCI test metadata and artifacts."""
+
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, main_config: MainConfig, client: CircleCIClient):
+        """Initialize the CircleCIScraper.
+
+        Args:
+            main_config (MainConfig): Directory information to store test results.
+            client (CircleCIClient): The CircleCI client to interact with the API.
+
+        """
+        self._client = client
+        self._test_result_dir = main_config.test_result_dir
+        self._test_metadata_dir = main_config.test_metadata_dir
+        self._test_artifact_dir = main_config.test_artifact_dir
+
+    def export_test_metadata_and_artifacts(
+        self,
+        pipeline_configs: list[CircleCIScraperPipelineConfig],
+        date_limit: datetime | None = None,
+    ) -> None:
+        """Export test metadata and artifacts for a list of pipelines.
+
+        Args:
+            pipeline_configs (List[CircleCIScraperPipelineConfig]): A list of pipeline
```

Comment:
Not sure if it matters, L75 says `list[...]` lowercase, L81 here says `List[...]` uppercase.